File: src/preprocessing/chunker.py
"""
Document and code chunking utilities for ARBuilder.
"""

import re
from dataclasses import dataclass, field
from typing import Optional
import tiktoken


@dataclass
class Chunk:
    """Represents a chunk of text or code."""
    content: str
    chunk_index: int
    total_chunks: int
    token_count: int
    metadata: dict = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "content": self.content,
            "chunk_index": self.chunk_index,
            "total_chunks": self.total_chunks,
            "token_count": self.token_count,
            **self.metadata,
        }


class DocumentChunker:
    """
    Chunk documents (markdown, text) for RAG ingestion.
    Uses semantic chunking based on headers and paragraphs.
    """

    def __init__(
        self,
        max_tokens: int = 512,
        overlap_tokens: int = 50,
        model: str = "cl100k_base",
    ):
        """
        Initialize the document chunker.

        Args:
            max_tokens: Maximum tokens per chunk.
            overlap_tokens: Token overlap between chunks.
            model: Tiktoken encoding model.
        """
        self.max_tokens = max_tokens
        self.overlap_tokens = overlap_tokens
        self.encoding = tiktoken.get_encoding(model)

    def count_tokens(self, text: str) -> int:
        """Count tokens in text."""
        return len(self.encoding.encode(text))

    def chunk(
        self,
        text: str,
        metadata: Optional[dict] = None,
    ) -> list[Chunk]:
        """
        Chunk a document into smaller pieces.

        Args:
            text: Document text to chunk.
            metadata: Metadata to attach to each chunk.

        Returns:
            List of Chunk objects.
        """
        if not text.strip():
            return []

        metadata = metadata or {}
        chunks = []

        # Split by headers first (semantic sections)
        sections = self._split_by_headers(text)

        current_chunk = ""
        current_tokens = 0

        for section in sections:
            section_tokens = self.count_tokens(section)

            # If section fits in current chunk
            if current_tokens + section_tokens <= self.max_tokens:
                current_chunk += section + "\n\n"
                current_tokens += section_tokens
            else:
                # Save current chunk if not empty
                if current_chunk.strip():
                    chunks.append(current_chunk.strip())

                # If section itself is too large, split by paragraphs
                if section_tokens > self.max_tokens:
                    sub_chunks = self._split_large_section(section)
                    chunks.extend(sub_chunks)
                    current_chunk = ""
                    current_tokens = 0
                else:
                    current_chunk = section + "\n\n"
                    current_tokens = section_tokens

        # Don't forget the last chunk
        if current_chunk.strip():
            chunks.append(current_chunk.strip())

        # Apply overlap and create Chunk objects
        final_chunks = self._apply_overlap(chunks)

        return [
            Chunk(
                content=chunk,
                chunk_index=i,
                total_chunks=len(final_chunks),
                token_count=self.count_tokens(chunk),
                metadata=metadata.copy(),
            )
            for i, chunk in enumerate(final_chunks)
        ]

    def _split_by_headers(self, text: str) -> list[str]:
        """Split text by markdown headers."""
        # Pattern matches markdown headers
        pattern = r"(?=^#{1,6}\s)"
        sections = re.split(pattern, text, flags=re.MULTILINE)
        return [s.strip() for s in sections if s.strip()]

    def _split_large_section(self, section: str) -> list[str]:
        """Split a large section by paragraphs."""
        paragraphs = section.split("\n\n")
        chunks = []
        current_chunk = ""
        current_tokens = 0

        for para in paragraphs:
            para_tokens = self.count_tokens(para)

            if current_tokens + para_tokens <= self.max_tokens:
                current_chunk += para + "\n\n"
                current_tokens += para_tokens
            else:
                if current_chunk.strip():
                    chunks.append(current_chunk.strip())

                # If paragraph is still too large, split by sentences
                if para_tokens > self.max_tokens:
                    sentence_chunks = self._split_by_sentences(para)
                    chunks.extend(sentence_chunks)
                    current_chunk = ""
                    current_tokens = 0
                else:
                    current_chunk = para + "\n\n"
                    current_tokens = para_tokens

        if current_chunk.strip():
            chunks.append(current_chunk.strip())

        return chunks

    def _split_by_sentences(self, text: str) -> list[str]:
        """Split text by sentences."""
        # Simple sentence splitting
        sentences = re.split(r"(?<=[.!?])\s+", text)
        chunks = []
        current_chunk = ""
        current_tokens = 0

        for sentence in sentences:
            sentence_tokens = self.count_tokens(sentence)

            if current_tokens + sentence_tokens <= self.max_tokens:
                current_chunk += sentence + " "
                current_tokens += sentence_tokens
            else:
                if current_chunk.strip():
                    chunks.append(current_chunk.strip())
                current_chunk = sentence + " "
                current_tokens = sentence_tokens

        if current_chunk.strip():
            chunks.append(current_chunk.strip())

        return chunks

    def _apply_overlap(self, chunks: list[str]) -> list[str]:
        """Apply overlap between chunks."""
        if len(chunks) <= 1 or self.overlap_tokens == 0:
            return chunks

        result = [chunks[0]]

        for i in range(1, len(chunks)):
            prev_chunk = chunks[i - 1]
            current_chunk = chunks[i]

            # Get overlap from end of previous chunk
            prev_tokens = self.encoding.encode(prev_chunk)
            overlap_text = self.encoding.decode(prev_tokens[-self.overlap_tokens:])

            # Prepend overlap to current chunk
            result.append(f"...{overlap_text}\n\n{current_chunk}")

        return result


class CodeChunker:
    """
    Chunk code files for RAG ingestion.
    Preserves function/struct boundaries where possible.
    """

    # Language-specific patterns for code splitting
    SPLIT_PATTERNS = {
        ".rs": [
            r"(?=^(?:pub\s+)?fn\s+\w+)",  # Rust functions
            r"(?=^(?:pub\s+)?struct\s+\w+)",  # Rust structs
            r"(?=^(?:pub\s+)?enum\s+\w+)",  # Rust enums
            r"(?=^(?:pub\s+)?impl\s+)",  # Rust impl blocks
            r"(?=^(?:pub\s+)?trait\s+\w+)",  # Rust traits
            r"(?=^(?:pub\s+)?mod\s+\w+)",  # Rust modules
        ],
        ".ts": [
            r"(?=^(?:export\s+)?(?:async\s+)?function\s+\w+)",  # TS functions
            r"(?=^(?:export\s+)?class\s+\w+)",  # TS classes
            r"(?=^(?:export\s+)?interface\s+\w+)",  # TS interfaces
            r"(?=^(?:export\s+)?type\s+\w+)",  # TS types
            r"(?=^(?:export\s+)?const\s+\w+\s*=\s*(?:async\s+)?\()",  # Arrow functions
        ],
        ".tsx": [
            r"(?=^(?:export\s+)?(?:default\s+)?function\s+\w+)",  # React components
            r"(?=^(?:export\s+)?const\s+\w+\s*[=:]\s*(?:\(\s*\)|React\.FC|FC))",  # FC components
            r"(?=^(?:export\s+)?const\s+use\w+)",  # Custom hooks
            r"(?=^(?:export\s+)?interface\s+\w+)",  # TS interfaces
            r"(?=^(?:export\s+)?type\s+\w+)",  # TS types
        ],
        ".jsx": [
            r"(?=^(?:export\s+)?(?:default\s+)?function\s+\w+)",  # React components
            r"(?=^(?:export\s+)?const\s+\w+\s*=\s*\()",  # Arrow function components
        ],
        ".js": [
            r"(?=^(?:export\s+)?(?:async\s+)?function\s+\w+)",  # JS functions
            r"(?=^(?:export\s+)?class\s+\w+)",  # JS classes
        ],
        ".sol": [
            r"(?=^contract\s+\w+)",  # Solidity contracts
            r"(?=^\s*function\s+\w+)",  # Solidity functions
            r"(?=^interface\s+\w+)",  # Solidity interfaces
            r"(?=^library\s+\w+)",  # Solidity libraries
        ],
        ".graphql": [
            r"(?=^type\s+\w+)",  # GraphQL types
            r"(?=^input\s+\w+)",  # GraphQL inputs
            r"(?=^enum\s+\w+)",  # GraphQL enums
            r"(?=^interface\s+\w+)",  # GraphQL interfaces
            r"(?=^query\s+\w+)",  # GraphQL queries
            r"(?=^mutation\s+\w+)",  # GraphQL mutations
        ],
        ".yaml": [
            r"(?=^[a-zA-Z_]+:(?:\s|$))",  # Top-level YAML keys
        ],
        ".yml": [
            r"(?=^[a-zA-Z_]+:(?:\s|$))",  # Top-level YAML keys
        ],
    }

    # File type to category mapping for metadata
    FILE_CATEGORIES = {
        ".rs": "rust_code",
        ".ts": "typescript_code",
        ".tsx": "react_component",
        ".jsx": "react_component",
        ".js": "javascript_code",
        ".sol": "solidity_contract",
        ".graphql": "graphql_schema",
        ".yaml": "config_yaml",
        ".yml": "config_yaml",
    }

    def __init__(
        self,
        max_tokens: int = 1024,
        overlap_lines: int = 5,
        model: str = "cl100k_base",
    ):
        """
        Initialize the code chunker.

        Args:
            max_tokens: Maximum tokens per chunk.
            overlap_lines: Line overlap between chunks.
            model: Tiktoken encoding model.
        """
        self.max_tokens = max_tokens
        self.overlap_lines = overlap_lines
        self.encoding = tiktoken.get_encoding(model)

    def count_tokens(self, text: str) -> int:
        """Count tokens in text."""
        return len(self.encoding.encode(text))

    def chunk(
        self,
        code: str,
        extension: str,
        metadata: Optional[dict] = None,
    ) -> list[Chunk]:
        """
        Chunk code into smaller pieces.

        Args:
            code: Code content to chunk.
            extension: File extension (e.g., ".rs", ".ts").
            metadata: Metadata to attach to each chunk.

        Returns:
            List of Chunk objects.
        """
        if not code.strip():
            return []

        metadata = metadata or {}
        metadata["language"] = extension.lstrip(".")
        metadata["file_category"] = self.FILE_CATEGORIES.get(extension, "code")

        # Try semantic splitting first
        chunks = self._semantic_split(code, extension)

        # If semantic split didn't work well, fall back to line-based
        if len(chunks) == 1 and self.count_tokens(chunks[0]) > self.max_tokens:
            chunks = self._line_split(code)

        # Further split any chunks that are too large
        final_chunks = []
        for chunk in chunks:
            if self.count_tokens(chunk) > self.max_tokens:
                final_chunks.extend(self._line_split(chunk))
            else:
                final_chunks.append(chunk)

        # Apply overlap
        final_chunks = self._apply_line_overlap(final_chunks)

        return [
            Chunk(
                content=chunk,
                chunk_index=i,
                total_chunks=len(final_chunks),
                token_count=self.count_tokens(chunk),
                metadata=metadata.copy(),
            )
            for i, chunk in enumerate(final_chunks)
        ]

    def _semantic_split(self, code: str, extension: str) -> list[str]:
        """Split code by semantic boundaries (functions, classes, etc.)."""
        patterns = self.SPLIT_PATTERNS.get(extension, [])

        if not patterns:
            return [code]

        # Combine patterns
        combined_pattern = "|".join(patterns)

        try:
            sections = re.split(combined_pattern, code, flags=re.MULTILINE)
            return [s.strip() for s in sections if s.strip()]
        except re.error:
            return [code]

    def _line_split(self, code: str) -> list[str]:
        """Split code by lines when semantic splitting isn't possible."""
        lines = code.splitlines()
        chunks = []
        current_chunk_lines = []
        current_tokens = 0

        for line in lines:
            line_tokens = self.count_tokens(line)

            if current_tokens + line_tokens <= self.max_tokens:
                current_chunk_lines.append(line)
                current_tokens += line_tokens
            else:
                if current_chunk_lines:
                    chunks.append("\n".join(current_chunk_lines))
                current_chunk_lines = [line]
                current_tokens = line_tokens

        if current_chunk_lines:
            chunks.append("\n".join(current_chunk_lines))

        return chunks

    def _apply_line_overlap(self, chunks: list[str]) -> list[str]:
        """Apply line-based overlap between chunks."""
        if len(chunks) <= 1 or self.overlap_lines == 0:
            return chunks

        result = [chunks[0]]

        for i in range(1, len(chunks)):
            prev_lines = chunks[i - 1].splitlines()
            overlap = prev_lines[-self.overlap_lines:]
            overlap_text = "\n".join(overlap)

            result.append(f"// ... continued from above\n{overlap_text}\n{chunks[i]}")

        return result

File: apps/web/src/lib/chunker.ts
/**
 * Chunking engine for Worker-native ingestion.
 *
 * TypeScript port of src/preprocessing/chunker.py.
 * Uses character-based token estimation (~4 chars/token) to avoid tiktoken.
 */

import type { GitHubFile } from "./github";
import { sha256 } from "./scraper";

export interface ProcessedChunk {
  id: string;
  content: string;
  chunk_index: number;
  source: string;
  url: string;
  title: string;
  category: string;
}

interface ChunkMetadata {
  url: string;
  title: string;
  category: string;
  source: "documentation" | "github";
}

// Language-specific split patterns (same as Python CodeChunker)
const CODE_SPLIT_PATTERNS: Record<string, RegExp[]> = {
  ".rs": [
    /(?=^(?:pub\s+)?fn\s+\w+)/m,
    /(?=^(?:pub\s+)?struct\s+\w+)/m,
    /(?=^(?:pub\s+)?enum\s+\w+)/m,
    /(?=^(?:pub\s+)?impl\s+)/m,
    /(?=^(?:pub\s+)?trait\s+\w+)/m,
    /(?=^(?:pub\s+)?mod\s+\w+)/m,
  ],
  ".ts": [
    /(?=^(?:export\s+)?(?:async\s+)?function\s+\w+)/m,
    /(?=^(?:export\s+)?class\s+\w+)/m,
    /(?=^(?:export\s+)?interface\s+\w+)/m,
    /(?=^(?:export\s+)?type\s+\w+)/m,
  ],
  ".js": [
    /(?=^(?:export\s+)?(?:async\s+)?function\s+\w+)/m,
    /(?=^(?:export\s+)?class\s+\w+)/m,
  ],
  ".sol": [
    /(?=^contract\s+\w+)/m,
    /(?=^\s*function\s+\w+)/m,
  ],
};

/**
 * Estimate token count using character approximation.
 * ~4 chars per token for English technical text, within ~10% accuracy.
 */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}

/**
 * Generate a deterministic chunk ID matching the Python pattern.
 * chunk_{sha256(url + content[:500])[:12]}
 */
export async function generateChunkId(
  url: string,
  content: string
): Promise<string> {
  const hashInput = url + content.slice(0, 500);
  const hash = await sha256(hashInput);
  return `chunk_${hash.slice(0, 12)}`;
}

/**
 * Chunk a documentation page into ProcessedChunks.
 * Port of Python DocumentChunker: split by headers -> paragraphs -> sentences.
 */
export async function chunkDocument(
  content: string,
  metadata: ChunkMetadata,
  maxTokens = 512,
  overlapTokens = 50
): Promise<ProcessedChunk[]> {
  if (!content.trim()) return [];

  // Split by markdown headers
  const sections = splitByHeaders(content);

  // Accumulate sections into chunks
  const rawChunks: string[] = [];
  let currentChunk = "";
  let currentTokens = 0;

  for (const section of sections) {
    const sectionTokens = estimateTokens(section);

    if (currentTokens + sectionTokens <= maxTokens) {
      currentChunk += section + "\n\n";
      currentTokens += sectionTokens;
    } else {
      if (currentChunk.trim()) {
        rawChunks.push(currentChunk.trim());
      }

      if (sectionTokens > maxTokens) {
        // Split large section by paragraphs
        const subChunks = splitLargeSection(section, maxTokens);
        rawChunks.push(...subChunks);
        currentChunk = "";
        currentTokens = 0;
      } else {
        currentChunk = section + "\n\n";
        currentTokens = sectionTokens;
      }
    }
  }

  if (currentChunk.trim()) {
    rawChunks.push(currentChunk.trim());
  }

  // Apply overlap
  const overlapped = applyOverlap(rawChunks, overlapTokens);

  // Create ProcessedChunk objects with IDs
  const chunks: ProcessedChunk[] = [];
  for (let i = 0; i < overlapped.length; i++) {
    const id = await generateChunkId(metadata.url, overlapped[i]);
    chunks.push({
      id,
      content: overlapped[i],
      chunk_index: i,
      source: metadata.source,
      url: metadata.url,
      title: metadata.title,
      category: metadata.category,
    });
  }

  return chunks;
}

/**
 * Chunk code files into ProcessedChunks.
 * Port of Python CodeChunker: split by language patterns -> line-based fallback.
 */
export async function chunkCode(
  files: GitHubFile[],
  metadata: Omit<ChunkMetadata, "title">,
  maxTokens = 1024,
  overlapLines = 5
): Promise<ProcessedChunk[]> {
  const chunks: ProcessedChunk[] = [];
  const repoName = metadata.url.replace(/\/$/, "").split("/").pop() || "";

  for (const file of files) {
    if (!file.content.trim()) continue;

    let rawChunks: string[];

    if (file.extension === ".md") {
      // Markdown files use document chunking logic
      const sections = splitByHeaders(file.content);
      rawChunks = accumulateSections(sections, 512);
      rawChunks = applyOverlap(rawChunks, 50);
    } else {
      // Code files use semantic splitting
      rawChunks = semanticSplitCode(file.content, file.extension);

      // If semantic split produced a single oversized chunk, fall back to lines
      if (rawChunks.length === 1 && estimateTokens(rawChunks[0]) > maxTokens) {
        rawChunks = lineSplit(file.content, maxTokens);
      }

      // Further split any oversized chunks
      const finalChunks: string[] = [];
      for (const chunk of rawChunks) {
        if (estimateTokens(chunk) > maxTokens) {
          finalChunks.push(...lineSplit(chunk, maxTokens));
        } else {
          finalChunks.push(chunk);
        }
      }
      rawChunks = finalChunks;

      // Apply line overlap
      rawChunks = applyLineOverlap(rawChunks, overlapLines);
    }

    // Create ProcessedChunk objects
    const title = `${repoName}/${file.path}`;
    for (let i = 0; i < rawChunks.length; i++) {
      const hashUrl = `${metadata.url}/${file.path}`;
      const id = await generateChunkId(hashUrl, rawChunks[i]);
      chunks.push({
        id,
        content: rawChunks[i],
        chunk_index: i,
        source: metadata.source,
        url: metadata.url,
        title,
        category: metadata.category,
      });
    }
  }

  return chunks;
}

// --- Internal helpers ---

function splitByHeaders(text: string): string[] {
  const sections = text.split(/(?=^#{1,6}\s)/m);
  return sections.map((s) => s.trim()).filter((s) => s.length > 0);
}

function splitLargeSection(section: string, maxTokens: number): string[] {
  const paragraphs = section.split("\n\n");
  const chunks: string[] = [];
  let currentChunk = "";
  let currentTokens = 0;

  for (const para of paragraphs) {
    const paraTokens = estimateTokens(para);

    if (currentTokens + paraTokens <= maxTokens) {
      currentChunk += para + "\n\n";
      currentTokens += paraTokens;
    } else {
      if (currentChunk.trim()) {
        chunks.push(currentChunk.trim());
      }

      if (paraTokens > maxTokens) {
        // Split by sentences
        chunks.push(...splitBySentences(para, maxTokens));
        currentChunk = "";
        currentTokens = 0;
      } else {
        currentChunk = para + "\n\n";
        currentTokens = paraTokens;
      }
    }
  }

  if (currentChunk.trim()) {
    chunks.push(currentChunk.trim());
  }

  return chunks;
}

function splitBySentences(text: string, maxTokens: number): string[] {
  const sentences = text.split(/(?<=[.!?])\s+/);
  const chunks: string[] = [];
  let currentChunk = "";
  let currentTokens = 0;

  for (const sentence of sentences) {
    const sentenceTokens = estimateTokens(sentence);

    if (currentTokens + sentenceTokens <= maxTokens) {
      currentChunk += sentence + " ";
      currentTokens += sentenceTokens;
    } else {
      if (currentChunk.trim()) {
        chunks.push(currentChunk.trim());
      }
      currentChunk = sentence + " ";
      currentTokens = sentenceTokens;
    }
  }

  if (currentChunk.trim()) {
    chunks.push(currentChunk.trim());
  }

  return chunks;
}

function accumulateSections(
  sections: string[],
  maxTokens: number
): string[] {
  const chunks: string[] = [];
  let currentChunk = "";
  let currentTokens = 0;

  for (const section of sections) {
    const sectionTokens = estimateTokens(section);
    if (currentTokens + sectionTokens <= maxTokens) {
      currentChunk += section + "\n\n";
      currentTokens += sectionTokens;
    } else {
      if (currentChunk.trim()) {
        chunks.push(currentChunk.trim());
      }
      if (sectionTokens > maxTokens) {
        chunks.push(...splitLargeSection(section, maxTokens));
        currentChunk = "";
        currentTokens = 0;
      } else {
        currentChunk = section + "\n\n";
        currentTokens = sectionTokens;
      }
    }
  }

  if (currentChunk.trim()) {
    chunks.push(currentChunk.trim());
  }

  return chunks;
}

function applyOverlap(chunks: string[], overlapTokens: number): string[] {
  if (chunks.length <= 1 || overlapTokens === 0) return chunks;

  const result = [chunks[0]];

  for (let i = 1; i < chunks.length; i++) {
    const prevChunk = chunks[i - 1];
    // Approximate overlap in characters
    const overlapChars = overlapTokens * 4;
    const overlapText = prevChunk.slice(-overlapChars);

    result.push(`...${overlapText}\n\n${chunks[i]}`);
  }

  return result;
}

function semanticSplitCode(code: string, extension: string): string[] {
  const patterns = CODE_SPLIT_PATTERNS[extension];
  if (!patterns || patterns.length === 0) return [code];

  // Combine patterns into a single regex
  const combined = new RegExp(
    patterns.map((p) => p.source).join("|"),
    "m"
  );

  const sections = code.split(combined);
  return sections.map((s) => s.trim()).filter((s) => s.length > 0);
}

function lineSplit(code: string, maxTokens: number): string[] {
  const lines = code.split("\n");
  const chunks: string[] = [];
  let currentLines: string[] = [];
  let currentTokens = 0;

  for (const line of lines) {
    const lineTokens = estimateTokens(line);

    if (currentTokens + lineTokens <= maxTokens) {
      currentLines.push(line);
      currentTokens += lineTokens;
    } else {
      if (currentLines.length > 0) {
        chunks.push(currentLines.join("\n"));
      }
      currentLines = [line];
      currentTokens = lineTokens;
    }
  }

  if (currentLines.length > 0) {
    chunks.push(currentLines.join("\n"));
  }

  return chunks;
}

function applyLineOverlap(chunks: string[], overlapLines: number): string[] {
  if (chunks.length <= 1 || overlapLines === 0) return chunks;

  const result = [chunks[0]];

  for (let i = 1; i < chunks.length; i++) {
    const prevLines = chunks[i - 1].split("\n");
    const overlap = prevLines.slice(-overlapLines);
    const overlapText = overlap.join("\n");

    result.push(`// ... continued from above\n${overlapText}\n${chunks[i]}`);
  }

  return result;
}
